libraries/AP_Quicktune/AP_Quicktune.cpp

```diff
@@ -0,0 +1,730 @@
+            return "Yaw I";
+        case Param::YAW_D:
+            return "Yaw D";
+        default:
+            INTERNAL_ERROR(AP_InternalError::error_t::flow_of_control);
+            return "UNK";
+    }
+}
+
+float AP_Quicktune::get_gain_mul()
+{
+   return expf(logf(2.0)/(UPDATE_RATE_HZ*double_time));
+}
+
+void AP_Quicktune::restore_all_params()
+{
+    for (int8_t pname = 0; pname < uint8_t(Param::END); pname++) {
+        if (BIT_IS_SET(param_changed, pname)) {
+            set_param_value(Param(pname), param_saved[pname]);
+            BIT_CLEAR(param_changed, pname);
+        }
+    }
+}
+
+void AP_Quicktune::save_all_params()
+{
+    // for pname in pairs(params) do
+    for (int8_t pname = 0; pname < uint8_t(Param::END); pname++) {
+        if (BIT_IS_SET(param_changed, pname)) {
+            set_and_save_param_value(Param(pname), get_param_value(Param(pname)));
+            param_saved[pname] = get_param_value(Param(pname));
+            BIT_CLEAR(param_changed, pname);
+        }
+    }
+}
+
+AP_Quicktune::Param AP_Quicktune::get_pname(AP_Quicktune::AxisName axis, AP_Quicktune::Stage stage)
+{
+    switch (axis)
+    {
+        case AxisName::RLL:
+            switch (stage)
+            {
+                case Stage::P:
+                    return Param::RLL_P;
+                case Stage::D:
+                    return Param::RLL_D;
+                case Stage::FLTT:
+                    return Param::RLL_FLTT;
+                case Stage::FLTD:
+                    return Param::RLL_FLTD;
+                default:
+                    INTERNAL_ERROR(AP_InternalError::error_t::flow_of_control);
+                    return Param::END;
+            }
+        case AxisName::PIT:
+            switch (stage)
+            {
+                case Stage::P:
+                    return Param::PIT_P;
+                case Stage::D:
+                    return Param::PIT_D;
+                case Stage::FLTT:
+                    return Param::PIT_FLTT;
+                case Stage::FLTD:
+                    return Param::PIT_FLTD;
+                default:
+                    INTERNAL_ERROR(AP_InternalError::error_t::flow_of_control);
+                    return Param::END;
+            }
+        case AxisName::YAW:
+            switch (stage)
+            {
+                case Stage::P:
+                    return Param::YAW_P;
+                case Stage::D:
+                    return Param::YAW_D;
+                case Stage::FLTT:
+                    return Param::YAW_FLTT;
+                case Stage::FLTD:
+                    return Param::YAW_FLTD;
+                default:
+                    INTERNAL_ERROR(AP_InternalError::error_t::flow_of_control);
+                    return Param::END;
+            }
+        default:
+            INTERNAL_ERROR(AP_InternalError::error_t::flow_of_control);
+            return Param::END;
+    }
+}
+
+AP_Quicktune::Stage AP_Quicktune::get_stage(AP_Quicktune::Param param)
```

Comment:
I find this very confusing, why not store the current axis and the current stage? I think storring the current param is partly what has resulted in odd things like this function.

Reply:
we also loop over all params in a few places, so both approaches are fine I think, I've gotten rid of the long switches